py3status/modules/speedtest.py

```diff
@@ -0,0 +1,296 @@
+    {server_host}           server host, eg 'speedtest.telecom-paristech.fr:8080'
+    {server_id}             server id, eg '11977'
+    {server_lat}            server latitude, eg '48.8742'
+    {server_latency}        server latency, eg 8.265
+    {server_lon}            server longitude, eg 2.3470
+    {server_name}           server name, eg 'Paris'
+    {server_sponsor}        server sponsor, eg 'Télécom ParisTech'
+    {server_url}            server url, eg 'http://speedtest.telecom-paristech.fr/upload.php'
+    {upload}                upload rate to speedtest server, human readable
+    {upload_raw}            upload rate to speedtest server, for format comparation
+    {upload_unit}           unit used for upload, eg 'MB/s'
+
+Color thresholds:
+    format:
+        xxx: print a color based on the value of `xxx` placeholder
+
+The module will be triggered on clic only. Not at start.
+
+Note that all placeholders have a version prefixed by `previous_`, eg `previous_download`.
+This can be usefull to compare two runs.
+
+Requires:
+    speedtest-cli: Command line interface for testing internet bandwidth using speedtest.net
+
+@author Cyril Levis (@cyrinux)
+
+Examples:
+```
+# colored based on thresholds
+speedtest {
+    thresholds = {
+        "download": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+        "ping": [(200, "bad"), (150, "orange"), (100, "degraded"), (10, "good")],
+        "quality": [
+            ("ok", "good"),
+            ("bad", "degraded"),
+            ("faster", "good"),
+            ("slower", "degraded"),
+            ("faster", "good"),
+        ],
+        "upload": [(0, "bad"), (1024, "degraded"), (1024 * 1024, "good")],
+    }
+    format = '[\?color=ping {ping} ms] [\?color=upload Up {upload} {upload_unit}]'
+    format += ' [\?color=download Down {download} {download_unit}]'
+}
+
+# colored based on comparation between two last runs
+speedtest {
+    format = '[[\?if=download_raw<previous_download_raw&color=degraded Faster]'
+    format += '|[\?color=ok Slower]]'
+}
+
+# compare only download
+speedtest {
+    format = '[[\?if=quality=slower&color=download_raw ↓ {download} {download_unit}]'
+    format += '[\?if=quality=faster&color=download_raw ↑ {download} {download_unit}]'
+    format += '[\?if=quality=ok&color=download_raw -> {download} {download_unit}]'
+    format += '[\?if=quality=bad&color=download_raw x {download} {download_unit}]]'
+    format += '|speedtest'
+}
+```
+
+SAMPLE OUTPUT
+{'full_text': u'Speedtest Down 0.247 MB/s Up 0.453 MB/s'}
+"""
+
+from json import loads
+from numbers import Number
+
+STRING_NOT_INSTALLED = "not installed"
+MISSING_PLACEHOLDER = "{download} or {upload} placeholder required"
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    button_refresh = 2
+    button_share = None
+    format = (
+        u"Speedtest [Up {upload} {upload_unit}] [Down {download} {download_unit}]"
+    )
```

Comment:
Your `format` prints extra spaces on right side on start. Test all your examples in real config, not inside `speedtest.py`. In thresholds... `1024 * 1024` throws an error.

Put colors in this format too except the words. We'll play with colors. Find what works well here. I'm aiming to have modules with a simple, clean theme so users can just add what they want and it would still look good on the bar.. Follow nocolor default, secondary darkgray, then colorized information.

For your module, I'm visualizing default `Speedtest` + lowercase darkgray `ping`, `up`, `down` prefixes with colorized stats... like `loadavg`s detailed format or `lm_sensors` default format.

Reply:
ok will try